internal/batch/batch_processor.go

```diff
@@ -411,76 +445,108 @@ func (bp *batchProcessor) maskContext(ctx context.Context, msg *fftypes.Message,
 	// The combination of the topic and group is the context
 	contextHash := fftypes.HashResult(hashBuilder)
 
-	// Get the next nonce for this context - we're the authority in the nextwork on this,
-	// as we are the sender.
-	gc := &fftypes.Nonce{
-		Context: contextHash,
-		Group:   msg.Header.Group,
-		Topic:   topic,
-	}
-	err = bp.database.UpsertNonceNext(ctx, gc)
-	if err != nil {
-		return "", nil, err
-	}
-
 	// Now combine our sending identity, and this nonce, to produce the hash that should
 	// be expected by all members of the group as the next nonce from us on this topic.
 	// Note we use our identity DID (not signing key) for this.
 	hashBuilder.Write([]byte(msg.Header.Author))
+
+	// Our DB of nonces we own, is keyed off of the hash at this point.
+	// However, before v0.14.2 we didn't include the Author - so we need to pass the contextHash as a fallback.
+	nonceKeyHash := fftypes.HashResult(hashBuilder)
+	nonce, err := bp.getNextNonce(ctx, state, nonceKeyHash, contextHash)
+	if err != nil {
+		return "", nil, err
+	}
+
+	// Now we have the nonce, add that at the end of the hash to make it unqiue to this message
 	nonceBytes := make([]byte, 8)
-	binary.BigEndian.PutUint64(nonceBytes, uint64(gc.Nonce))
+	binary.BigEndian.PutUint64(nonceBytes, uint64(nonce))
 	hashBuilder.Write(nonceBytes)
 
 	pin := fftypes.HashResult(hashBuilder)
```

Comment:
Isn't this recomputing the same value from line 455?

Reply:
Line 455 is the state of the hash after we've written:
- The topic
- The group
- The author

Then afterwards is generating a big endian 8byte hex value for the nonce (which we've just determined from the above) and adding it to the end of the hash.

Example here:
https://github.com/hyperledger/firefly/blob/25a2b21893c5d32c117977db4ea1bb52aa3294ad/internal/batch/batch_manager_test.go#L189